### PR TITLE
add bits_to_target

### PR DESCRIPTION
**Pull request by harsh-ps-2003**
- [x] issue #9 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)
- [x] code change includes tests

Adds bits_to_target with a custom power method of u256, and uses it to validate_target

### CODE REVIEW

**Comment by maciejka** [comment]
Where are the test cases?

**Review by maciejka — COMMENTED** [review]
Where are the test cases?

**Comment by harsh-ps-2003** [comment]
> Where are the test cases?

Will be adding them later today most probably! Suddenly got busy with some work! Thanks for the patience

File: src/engine.cairo [hunk 1]
```diff
@@ -164,8 +168,54 @@ pub impl BlockHeaderEngineImpl of BlockHeaderEngineTrait {
...
+
+    if exponent == 0 {
+        // Special case: exponent 0 means we use the mantissa as-is
+        return Result::Ok(target);
+    } else if exponent <= 3 {
+        // For exponents 1, 2, and 3, divide by 256^(3 - exponent) i.e right shift
+        let shift = 8 * (3 - exponent);
+        let mut multiplier: u256 = 1.into();
+        let mut i = 0;
+        while i < shift {
```
Comment (maciejka) [inline src/engine.cairo:192]: why don't you just use `div` here?

File: src/engine.cairo [hunk 2]
```diff
@@ -164,8 +168,54 @@ pub impl BlockHeaderEngineImpl of BlockHeaderEngineTrait {
 }
 
 // Helper functions
-fn bits_to_target(bits: u32) -> u256 {
-    0
+pub fn bits_to_target(bits: u32) -> Result<u256, felt252> {
```
Comment (maciejka) [inline src/engine.cairo:171]: Implementation here: https://github.com/ZeroSync/ZeroSync/blob/f1bd4736b0bac190ac48cbc8d5fcd92e2e1be7b9/src/block/block_header.cairo#L178 looks much simpler. It seems there is no need to use loops.

File: tests/tests.cairo [hunk 3]
```diff
@@ -0,0 +1,72 @@
+use raito::engine::bits_to_target;
```
Comment (maciejka) [inline tests/tests.cairo:1]: Use test cases from: https://github.com/ZeroSync/ZeroSync/blob/main/tests/unit/block/test_bits_and_target.json

**Review by maciejka — CHANGES_REQUESTED** [review]
Please squash prs #15 and #16.

**Comment by maciejka** [comment]
How are you? Is there any progress? There are other issues waiting for this one to finish...

**Comment by harsh-ps-2003** [comment]
Will pick this up today again! Thanks for the patience

**Comment by harsh-ps-2003** [comment]
Can you squash merge the #15 and #16 ?

**Comment by harsh-ps-2003** [comment]
@maciejka resolving the conflicts is a mess as the project changed quite a lot! can I close both PRs and open 1 new one with all the changes instead!

**Review by maciejka — APPROVED** [review]

**Comment by maciejka** [comment]
Fix fmt and tests, please.

**Comment by maciejka** [comment]
> @maciejka suddenly I get error
> 
> ```
> Error: `scarb metadata` exited with error
> 
> stdout:
> error: failed to lookup for `cairo_test ^2.7.0` in registry: registry+https://there-is-no-default-registry-yet.com/
> ```
> 
> when trying to run `scarb test` locally!

Please ask on scarb tg channel: https://t.me/+1pMLtrNj5NthZWJk

File: src/utils.cairo [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+// Bitwise shift right for u256
+pub fn shr(value: u256, shift: u32) -> u256 {
+    value / fast_pow(2.into(), shift.into())
+}
+
+// Fast exponentiation using the square-and-multiply algorithm
+// Reference: https://github.com/keep-starknet-strange/alexandria/blob/bcdca70afdf59c9976148e95cebad5cf63d75a7f/packages/math/src/fast_power.cairo#L12
+pub fn fast_pow(base: u256, exp: u32) -> u256 {
+    if exp == 0 {
+        return 1.into();
```
Comment (maciejka) [inline src/utils.cairo:18]: just 1_u256

File: src/utils.cairo [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    let mut res: u256 = 1.into();
```
Comment (maciejka) [inline src/utils.cairo:21]: same here

**Review by maciejka — CHANGES_REQUESTED** [review]

**Review by maciejka — APPROVED** [review]

**Comment by maciejka** [comment]
@harsh-ps-2003 please fix conflicts

**Comment by harsh-ps-2003** [comment]
Thanks for the patience!

**Pull request merged**